docs/okeps/okep-5094-layer2-transit-router.md

```diff
@@ -0,0 +1,991 @@
+
+```mermaid
+%%{init: {"nodeSpacing": 20, "rankSpacing": 100}}%%
+flowchart TD
+    classDef nodeStyle fill:orange,stroke:none,rx:10px,ry:10px,font-size:25px;
+    classDef vmStyle fill:blue,stroke:none,color:white,rx:10px,ry:10px,font-size:25px;
+    classDef portStyle fill:#3CB371,color:black,stroke:none,rx:10px,ry:10px,font-size:25px;
+    classDef routerStyle fill:brown,color:white,stroke:none,rx:10px,ry:10px,font-size:25px;
+    classDef switchStyle fill:brown,color:white,stroke:none,rx:10px,ry:10px,font-size:25px;
+    classDef termStyle font-family:monospace,fill:black,stroke:none,color:white;
+    subgraph node2
+        subgraph GR-node2
+            rtos-GR-node2["rtos-GR-node2 100.65.0.3/16 (0a:58:64:41:00:03)"]
+        end
+        subgraph layer2-switch_node2["layer2-switch"]
+            stor-GR-node1["stor-GR-node1
+            type: <b>remote</b><br>requested-tnl-key: <b>4</b>"]
+            stor-GR-node2["stor-GR-node2
+            type: <b>router</b><br>requested-tnl-key: <b>5</b>"]
+        end
+    end
+    subgraph node1["node1"]
+        subgraph GR-node1
+            rtotr-GR-node1["trtor-GR-node1
+            100.65.0.2/16 100.88.0.5/31 (0a:58:64:41:00:02)"]
+        end
+        subgraph transit_router_node1["transit_router "]
+            trtor-GR-node1["trtor-GR-node1 100.88.0.4/31"]
+            rtos-layer2-switch["rtos-layer2-switch 203.203.0.1 (0a:58:CB:CB:00:01)"]
+        end
+        subgraph layer2-switch_node1["layer2-switch"]
+            stor-transit_router["stor-transit_router
+            type: <b>router</b><br>requested-tnl-key: <b>4</b>"]
+            stor-GR-node1-node2["stor-GR-node2
+            type: <b>remote</b><br>requested-tnl-key: <b>5</b>"]
+        end
+    end
+
+
+
+    rtotr-GR-node1 <--> trtor-GR-node1
+    rtos-layer2-switch <--> stor-transit_router
+    stor-GR-node1 <--> stor-transit_router
+    stor-GR-node2 <--> rtos-GR-node2
+    stor-GR-node1-node2 <--> stor-GR-node2
+
+    class VM vmStyle;
+    class rtotr-GR-node1 portStyle;
+    class rtos-GR-node2 portStyle;
+    class stor-GR-node2 portStyle;
+    class stor-GR-node1-node2 portStyle;
+    class stor-GR-node1 portStyle;
+    class trtor-GR-node1 portStyle;
+    class trtor-GR-node2 portStyle;
+    class stor-transit_router portStyle;
+    class rtos-layer2-switch portStyle;
+    class GR-node1 routerStyle;
+    class GR-node2 routerStyle;
+    class transit_router_node1 routerStyle;
+    class layer2-switch_node1 switchStyle;
+    class layer2-switch_node2 switchStyle;
+    class term termStyle;
+    class node1,node2 nodeStyle;
+```
+
+### Egress IP changes (TBD)
+
+### Transit subnet conflict
+
+Before this change Layer2 networks using Subnet overlapping with the `transitSubnet` were allowed to be created (and would be working just fine).
+Now it is not possible anymore, and we need to decide what to do in this case:
+1. Before upgrading to the new topology, check that no Layer2 networks with overlapping subnet are present.
```

Comment:
This needs more details. How exactly does this work? Are you going to put the UDNs that are a problem into error state? Are you going to refuse to annotate k8s.ovn.org/layer2-topology-version: 2.0 due to this error?

Reply:
there is no way to put networks in error state, because we use per-node annotation now
so I just won't upgrade the node if layer2 pods are present

Reply:
what would be the netmask of this automatically generated subnet ? how would OVN-K infer it ?

Reply:
we can do some subnet math to figure out a new subnet that is the same mask and doesnt collide with join/pod subnet. I don't think we need to detail that out here.

Reply:
that isn't practical because what if someone chose a pod subnet of 100.0.0.0/8? You have to calculate based on the mask itself to find the a non overlapping subnet.

Reply:
not like you have to, the difference will be around 100ns
but I was describing the idea more than the actual implementation to achieve that